geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/KryoLazyDensityFilter.java

```diff
@@ -0,0 +1,379 @@
+    private final String GRID_HEIGHT_OPT = "grid_height";
+    private final String WEIGHT_OPT = "weight";
+    private final String GEOM_INDEX = "geom_index";
+    private final String WEIGHT_INDEX = "weight_index";
+
+    private int weightIndex = -2;
+    private int geomIndex = -1;
+
+    private GridSnap gridSnap = null;
+    private Map<String, Object> options = new HashMap<>();
+
+    private final static Logger logger = LoggerFactory.getLogger(KryoLazyDensityFilter.class);
+
+    public void configure(String sftString, Hints hints){
+        this.sftString = sftString;
+        configureSFT();
+        configureOptions(hints);
+    }
+
+    public void configure(SimpleFeatureType sft, Hints hints) {
+        this.sft = sft;
+        this.sftString = SimpleFeatureTypes.encodeType(sft, true);
+        configureOptions(hints);
+    }
+
+    public KryoLazyDensityFilter(String sftString, Hints hints) {
+        this.sftString = sftString;
+        configureSFT();
+        configureOptions(hints);
+    }
+
+    public KryoLazyDensityFilter(String sftString, Map<String, Object> options) {
+        this.sftString = sftString;
+        configureSFT();
+        if (options == null) {
+            return;
+        }
+        this.options = options;
+
+        this.weightIndex = (Integer) options.getOrDefault(WEIGHT_INDEX, -2);
+        this.geomIndex = (Integer) options.getOrDefault(GEOM_INDEX, -1);
+        Envelope envelope = (Envelope) options.getOrDefault(ENVELOPE_OPT, null);
+
+        if(envelope != null){
+            int width = (Integer) options.get(GRID_WIDTH_OPT);
+            int height = (Integer) options.get(GRID_HEIGHT_OPT);
+            this.gridSnap = new GridSnap(envelope, width, height);
+        }
+    }
+
+    public KryoLazyDensityFilter(SimpleFeatureType sft, Hints hints) {
+        this.sft = sft;
+        this.sftString = SimpleFeatureTypes.encodeType(sft, true);
+        configureOptions(hints);
+    }
+
+    private void configureOptions(Hints hints) {
+        geomIndex = getGeomIndex();
+        options.put(GEOM_INDEX, geomIndex);
+
+        if (hints == null) {
+            return;
+        }
+
+        Envelope envelope = getDensityEnvelope(hints);
+        int width = getDensityWidth(hints);
+        int height = getDensityHeight(hints);
+        String weight = getDensityWeight(hints);
+        options.put(WEIGHT_OPT, weight);
+        options.put(ENVELOPE_OPT, envelope);
+        options.put(GRID_WIDTH_OPT, width);
+        options.put(GRID_HEIGHT_OPT, height);
+        gridSnap = new GridSnap(envelope, width, height);
+        int index = sft.indexOf((String) options.get(WEIGHT_OPT));
+        weightIndex = index;
+        options.put(WEIGHT_INDEX, weightIndex);
+    }
+
+    private void configureSFT() {
+        sft = SimpleFeatureTypes.createType("densitySft", sftString);
+        serializer = new KryoFeatureSerializer(sft, SerializationOptions.withoutId());
```

Comment:
there are a lot of different constructors/configure methods - could we consolidate them down so that: there are as few as are actually needed, and: they call each other? this approach seems likely to end with some things not initiated correctly.